File: write_to_graphite.rb
#!/usr/bin/env ruby

#
# Writes a passed value to the graphite server of your choosing.
# * *Arguments:
#  host <h>
#  port<p>
#  tree(Structure to create in Graphite) <t>
#  value (to be written to graphite <v>
#
# Examples:
# +require './write_to_graphite'+
# +graphite = Write_To_Graphite::Graph.new('graphite.yourdomain.com', 2003, 'Data.To.Log')+
# +graphite.write_out(value)+
#
# @author Tim Hunter

require 'rubygems'
require 'socket'
require 'logger'

module Write_To_Graphite

  class Graph
    def initialize (h,p,t)
      @log = Logger.new('/var/log/write_out.log', 'daily') #//TODO: path should be dynamic based on OS so that this can be run from Win too
      @log.level = Logger::DEBUG
      @graphite_host = h
      @graphite_port = p
      time = Time.new
      @epoch = time.to_i #get me an epoch time to write to graphite
      @tree = t #Tree is the structure in graphite
    end #end Init

    def write_out(v) #v = passed value to be written to graphite
      value = v
      begin#open my socket
        socket = TCPSocket.open("#{@graphite_host}", "#{@graphite_port}") 
      rescue#log out errors and close the socket and log nicely
        @log.error "error #{$!}"
        socket.close
        @log.close
      else#write a log entry for this transaction and then write the data to graphite, then clean up after ourselves
        @log.info "writing #{@tree} #{value} #{@epoch} to #{@graphite_host}:#{@graphite_port}"
        socket.write("#{@tree} #{value} #{@epoch}\n") #write out to graphite
        socket.close
        @log.close
      end

    end #end write_out
 end #end Graph
end #end Module

File: write_to_graphite.py
#!/usr/bin/env python

#
# Writes a passed value to the graphite server of your choosing.
# * *Arguments:
#  host 
#  port
#  tree(Structure to create in Graphite) 
#  value (to be written to graphite 
#
# Examples:
# +from write_to_graphite import w2g+
# +w2g('graphite.yourdomain.com', 2003, 'Data.To.Log', value)+
# 
#
# @author Tim Hunter

import socket
import time
import argparse


def w2g(server, port, tree, value):
	##	define our graphpoint to send (tree, value, timestamp)
	graphpoint = "%s %s %d\n" % (tree, value, int(time.time()))
	##	define and opne our socket to the graphite host
	con = socket.socket()
	con.connect((server, port))
	##	send the graph point to the graphite server
	con.sendall(graphpoint)
	##	cleanup
	con.close()

if __name__ == '__main__':
	##	let us run this from the command line too
	parse = argparse.ArgumentParser(description="Graphite Writer")
	parse.add_argument('-s', '--server', type=str, help="Graphite Server name")
	parse.add_argument('-p', '--port', type=int, help="Graphite Server Port")
	parse.add_argument('-t', '--tree', type=str, help="Graphite tree Structure to write to. ex: Production.Server1.CPU")
	parse.add_argument('-v', '--value', type=str, help="Value to send to Graphite")
	##	sanity checks, make sure we have all the nessecary data
	args = parse.parse_args()
	if args.server and args.port and args.tree and args.value: 
		w2g(args.server,args.port,args.tree,args.value)
	else:
		parse.print_help()
